lib/core/src/Cardano/Wallet.hs

```diff
@@ -1976,21 +1977,62 @@ joinStakePool ctx currentEpoch knownPools pid poolStatus wid argGenChange pwd =
         let action = if isKeyReg then Join pid else RegisterKeyAndJoin pid
         liftIO $ traceWith tr $ MsgIsStakeKeyRegistered isKeyReg
 
-        selection <- withExceptT ErrJoinStakePoolSelectCoin $
+        cs <- withExceptT ErrJoinStakePoolSelectCoin $
             selectCoinsForDelegation @ctx @s @t @k ctx wid action
 
+        withRootKey @_ @s ctx wid pwd ErrJoinStakePoolRootKey $ \xprv scheme -> do
+            let pwdP = preparePassphrase scheme pwd
+            let rewardAcnt = getRawKey $ deriveRewardAccount @k pwdP xprv
+            let certs = (mkDelegationCertificates tl) (W.Join pid)
+                    (rewardAcnt, pwdP) cs
```

Comment:
We should _not_ require the passphrase nor the private key for this. Actually, the main use-case for such function is when used for hardware devices integration in which case we do not have access to the root private key whatsoever. 

It is okay to enforce that `s ~ SeqState n k` here because it makes no sense anyway to call this for a `RndState` (random wallets) which have no delegation capabilities. From there, it is possible to access the `rewardAccountKey` corresponding to that particular account without any need to access the root private key.

Reply:
We also need `k ~ ShelleyKey` afais, for `getKey`.

Reply:
You can use `WalletKey k` as a constraint and use `getRawKey`

Reply:
```
src/Cardano/Wallet/Jormungandr/Api/Server.hs:191:13: error:
    • Couldn't match type ‘JormungandrKey’
                     with ‘Cardano.Wallet.Primitive.AddressDerivation.Shelley.ShelleyKey’
        arising from a use of ‘joinStakePool’
    • In the first argument of ‘(:<|>)’, namely
        ‘joinStakePool
           jormungandr (knownStakePools spl) (getPoolLifeCycleStatus spl)’
      In the second argument of ‘(:<|>)’, namely
        ‘joinStakePool
           jormungandr (knownStakePools spl) (getPoolLifeCycleStatus spl)
           :<|> quitStakePool jormungandr :<|> delegationFee jormungandr’
      In the expression:
        (listPools spl)
          :<|>
            joinStakePool
              jormungandr (knownStakePools spl) (getPoolLifeCycleStatus spl)
              :<|> quitStakePool jormungandr :<|> delegationFee jormungandr
    |
191 |             joinStakePool jormungandr
```

Reply:
nvm, found it